gluten-core/src/main/scala/io/glutenproject/expression/ScalarSubqueryTransformer.scala

```diff
@@ -26,6 +26,10 @@ class ScalarSubqueryTransformer(plan: BaseSubqueryExec, exprId: ExprId,
   extends ExpressionTransformer {
 
   override def doTransform(args: java.lang.Object): ExpressionNode = {
+    // don't trigger collect when in validation phase
+    if (TransformerState.underValidationState) {
+      return ExpressionBuilder.makeLiteral(null, query.dataType, true)
+    }
```

Comment:
@rui-mo Aggreed. Method in `ExpressionBuilder.makeLiteral` has some validation logic  that rely on actual subquery values.

Reply:
Do you think it sounds good to add a config to control whether subquery is executed during validation? By default, the subquery validation still takes effect.

Reply:
I plan to skip execution only for data insensitive data Types. For data sensitive dataTypes like StructType, may be we can add a result cache or something similar to that(the idea is immature).

Reply:
This is OK to me if it can bypass this issue. Can we move the `valueSensitiveDataType` method into ScalarSubqueryTransformer? Because looks it helps handle a special case in this case, and is not generally used.

Reply:
> When scala subquery filters pushed down to scan by gluten, duplicate execution happens too. It's not in validation stage.

@loudongfeng Could you explain more about why duplication happens?

Reply:
This query can reproduce:
```sql
explain select i_item_sk from item where i_product_name = (select max(i_product_name) from item);
```
SparkSQL output
```sql
== Physical Plan ==
*(1) Project [i_item_sk#7L]
+- *(1) Filter (isnotnull(i_product_name#28) AND (i_product_name#28 = Subquery scalar-subquery#1, [id=#38]))
   :  +- Subquery scalar-subquery#1, [id=#38]
   :     +- SortAggregate(key=[], functions=[max(i_product_name#52)])
   :        +- Exchange SinglePartition, ENSURE_REQUIREMENTS, [plan_id=36]
   :           +- SortAggregate(key=[], functions=[partial_max(i_product_name#52)])
   :              +- *(1) ColumnarToRow
   :                 +- FileScan parquet tpcds_parquet.item[i_product_name#52] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/home/nemon/tpcd..., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<i_product_name:string>
   +- *(1) ColumnarToRow
      +- FileScan parquet tpcds_parquet.item[i_item_sk#7L,i_product_name#28] Batched: true, DataFilters: [isnotnull(i_product_name#28)], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/home/nemon/tpcd..., PartitionFilters: [], PushedFilters: [IsNotNull(i_product_name)], ReadSchema: struct<i_item_sk:bigint,i_product_name:string>

```
Gluten output
```sql
== Physical Plan ==
CHNativeColumnarToRow
+- ^(2) ProjectExecTransformer [i_item_sk#7L]
   +- ^(2) FilterExecTransformer (isnotnull(i_product_name#28) AND (i_product_name#28 = Subquery scalar-subquery#1, [id=#57]))
      :  +- Subquery scalar-subquery#1, [id=#57]
      :     +- SortAggregate(key=[], functions=[max(i_product_name#52)])
      :        +- CHNativeColumnarToRow
      :           +- CoalesceBatches
      :              +- ColumnarExchange SinglePartition, ENSURE_REQUIREMENTS, false, [plan_id=53], [id=#53], [OUTPUT] List(max:StringType)
      :                 +- RowToCHNativeColumnar
      :                    +- SortAggregate(key=[], functions=[partial_max(i_product_name#52)])
      :                       +- CHNativeColumnarToRow
      :                          +- ^(1) NativeFileScan parquet tpcds_parquet.item[i_product_name#52] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/home/nemon/tpcd..., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<i_product_name:string>
      +- ^(2) NativeFileScan parquet tpcds_parquet.item[i_item_sk#7L,i_product_name#28] Batched: true, DataFilters: [isnotnull(i_product_name#28), (i_product_name#28 = Subquery scalar-subquery#1, [id=#57])], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/home/nemon/tpcd..., PartitionFilters: [], PushedFilters: [IsNotNull(i_product_name)], ReadSchema: struct<i_item_sk:bigint,i_product_name:string>
            +- Subquery scalar-subquery#1, [id=#57]
               +- SortAggregate(key=[], functions=[max(i_product_name#52)])
                  +- CHNativeColumnarToRow
                     +- CoalesceBatches
                        +- ColumnarExchange SinglePartition, ENSURE_REQUIREMENTS, false, [plan_id=53], [id=#53], [OUTPUT] List(max:StringType)
                           +- RowToCHNativeColumnar
                              +- SortAggregate(key=[], functions=[partial_max(i_product_name#52)])
                                 +- CHNativeColumnarToRow
                                    +- ^(1) NativeFileScan parquet tpcds_parquet.item[i_product_name#52] Batched: true, DataFilters: [], Format: Parquet, Location: InMemoryFileIndex(1 paths)[file:/home/nemon/tpcd..., PartitionFilters: [], PushedFilters: [], ReadSchema: struct<i_product_name:string>

```
In Gluten plan, scalar-subquery#1 is pushed down to scan,so `ScalarSubqueryTransformer.doTransform` execute twice during whole stage transform.One for FilterExecTransformer, one for TableScanTransformer.
May be we can fix this in another PR.

Reply:
Thank you. Yes, we can do that.
There are some logics in Filter transformer to prevent duplicate execution for the filter conditions pushed down into Scan. Do you suggest although with this logic the Subquery is still executed twice?

https://github.com/oap-project/gluten/blob/main/gluten-data/src/main/scala/io/glutenproject/execution/GlutenFilterExecTransformer.scala#L39-L44

Reply:
Thanks for your information. I find that this duplicate execution only happens for ch backend + parquet file format. After changing this line from FilterExecTransformer to CHFilterExecTransformer, then no duplicate execution.
https://github.com/oap-project/gluten/blob/main/backends-clickhouse/src/main/scala/io/glutenproject/backendsapi/clickhouse/CHSparkPlanExecApi.scala#L92

@zzcclp Any idea why we use two different FilterExecTransformer for ch backend?

Reply:
I have created a seperate ticket to track the issue mentioned above. #1227 
Shall we merge this PR first? @rui-mo

Reply:
Sure. If there's no other comment.